### PR TITLE
Notarization through the Notary API

### LINKED ISSUE #591: [apple-codesign] Use new REST API for notarization
Heyo,

At WWDC22, Apple announced that they were releasing a new REST API for notarization. I believe this would allow getting rid of the dependency on Apple Transporter. It supports submitting a new package to the notarization service, and getting the submission status and logs.

- Documentation: https://developer.apple.com/documentation/notaryapi/submit_software
- WWDC Video: https://developer.apple.com/videos/play/wwdc2022/10109/

### PR DESCRIPTION
**Pull request by roblabla**
This MR adds notarization through the Notary API.

Fixes #591 

Currently, this is only implemented for code bundle signing. I need to implement pkg and dmg signing next, and do a lot of code cleanup.

This API is quite a bit faster than the old method. My code gets notarized in around 20 seconds with this API, whereas it took around a minute with the old Transporter API.

---

Note that I'm also having a very messy dependency hell issue. PyOxidizer depends on `pgp`, that fell in the pitfall of using non-standard dependency version requirements on zeroize... This means that it is currently impossible to use `pgp` and `aws-sdk-s3` in the same dependency graph. I have reported the problem to pgp, and have a patch to fix the issue. 

Hopefully `pgp` will release a new version to fix this issue.

### CODE REVIEW

**Comment by indygreg** [comment]
Thank you for submitting this!

Regarding the dependency hell issue, yes, this is something I've run into as well. `pgp` is using some old crate versions and is holding back upgrading a lot of dependencies in `apple-codesign` as well as other crates in this repo. It is super annoying. https://github.com/rpgp/rpgp/pull/165 should hopefully fix most issues. Although I wasn't aware of that non-standard pinning of the `zeroize` crate. Maybe your PR will be sufficient to unlock progress, as the `zeroize` crate is the one that breaks most frequently when I try to modernize crypto dependencies in this repo.

**Review by indygreg — COMMENTED** [review]
Thank you so much for this PR! Using the new notarization API and deleting the old transporter code should definitely be the marquee feature for the next release of this create and I can't wait to ship pure Rust notarization!

The approach so far in this PR seems very reasonable to me. I don't see any roadblocks other than the `pgp` crate patching. Regarding the `pgp` crate, ideally we'd wait for them to publish a new version. But I'm not opposed to shipping a version with a patched `pgp` crate as a temporary workaround. At the very least I'd like to see that crate accept a PR with a fix, as I would prefer to pin to an unreleased commit in the crate's canonical repo. I suppose we'll wait and see what happens with your PR.

Since this is marked as a work in progress, I guess ping me when you think this is ready for a real review. If you don't want to push this PR over the finish line, that's fine by me too: I can finish it the next time I sit down to hack on this project.

Thank you again for your work on this feature!

File: apple-codesign/src/notarization.rs [hunk 1]
```diff
@@ -656,12 +669,61 @@ impl Notarizer {
...
         wait_limit: Option<Duration>,
     ) -> Result<NotarizationUpload, AppleCodesignError> {
-        let (_temp_dir, dest_dir) = create_itmsp_temp_dir()?;
-        warn!("writing App Store Package to {}", dest_dir.display());
+        //warn!("writing App Store Package to {}", dest_dir.display());
 
-        write_bundle_to_app_store_package(bundle, &dest_dir)?;
+        let zipfile = bundle_to_zip(bundle)?;
+
+        let digest = sha2::Sha256::digest(&zipfile);
```
[the remark this answers is not in the document]
Comment (roblabla) [reply apple-codesign/src/notarization.rs:214]: Yup. That's what I did in my latest push.

**Comment by roblabla** [comment]
Alright so I should have gotten PKG and DMG in. I still need to actually test it works (only bundle was tested for now), and assuming it all works, make the commit history a bit nicer :^).

Another thing I'd like to improve is getting richer error information by extracting it from the submission logs. There's a lot of really useful information in there, but right now it's printed as an ugly json blob. I want to add some method to pretty print it.

**Comment by indygreg** [comment]
Looks like your zeroize PR merged in the rpgp repo! Hopefully that means a release is near. If not, I'm not opposed to throwing up a fork on crates.io as a workaround: I'm tired of working around dependency hell resulting from that crate.

**Comment by indygreg** [comment]
I was having my own go at upgrading dependencies and it looks like yubikey depending on rsa 0.5 is going to be a hurdle. I filed https://github.com/iqlusioninc/yubikey.rs/issues/393 to request a release.

**Comment by roblabla** [comment]
Yeah, I ran into yubikey-rs problems, and also rpm-rs problems. See https://github.com/roblabla/rpm-rs/tree/update-packages

**Comment by indygreg** [comment]
I haven't forgotten about this PR. I wanted to get the dependency upgrades merged first because they are a pain to update. As I was doing that, I discovered a dependency hell due to the `yubikey-rs` crate and wanted to get that resolved. It is kind of in limbo at the moment. I'll try to find time to take a look at this over the weekend.

**Review by indygreg — APPROVED** [review]
I incrementally applied this PR locally, crafting a few separate commits from your work. I retained your authorship of those commits.

I need to spend some time testing this and updating docs to reflect the new state. But at this point the code is pretty much locked in and will be pushed, hopefully some time this weekend. The PR should automatically close when I do so.

I sincerely thank you for this contribution. Pure Rust notarization was the last major missing feature for this crate and I'm excited it is finally implemented! I'll try to push out a new release with this major feature soon as well. (But there are some other known bugs with the crate that I may want to try to fix first.)

File: apple-codesign/src/app_store_connect.rs [hunk 2]
```diff
@@ -216,6 +216,156 @@ pub struct MoreInfo {
...
+    pub job_id: String,
+    pub log_format_version: u64,
+    pub sha256: String,
+    pub status: SubmissionResponseStatus,
+    pub status_code: u64,
+    pub status_summary: String,
+    #[serde(default)]
+    pub ticket_contents: Vec<TicketContent>,
+    pub upload_date: String,
+}
```
[the remark this answers is not in the document]
Comment (roblabla) [reply apple-codesign/src/app_store_connect.rs:367]: I just looked at what data the endpoint was returning.

[the remark this answers is not in the document]
Comment (roblabla) [reply apple-codesign/src/app_store_connect.rs:367]: Also another thing: I found out by trial and error that the correct S3 region to upload the data to is us-west-2, but this is not documented anywhere... I'm not sure how apple expects us to figure this out.

**Comment by indygreg** [comment]
As you can see, I pushed this. I also did a bit of refactoring after the push, mostly to clean up the UI around notarization.

The one part of your PR I did not preserve is the structs for parsing the log JSON. We don't need that yet, as we just print the JSON instead.

I did add your GitHub username and reported full name from your GitHub profile to the changelog. And I plan to say something similar in an upcoming blog post about the pending 0.17 release. If you would like me to withhold your username or full name or identify you some other way, just let me know and I'll respect your wishes. I'll likely publish 0.17 on Sunday and the blog post on Monday morning California time.

Thanks again for this amazing contribution!

**Pull request closed without merging**